### PR TITLE
Dynamic version selector for Npm and Yarn

### PR DESCRIPTION
**Pull request by sachin-sandhu**
### What are you trying to accomplish?

**Background:** Dependabot relies on ecosystem helpers (package managers) to perform dependencies updates. The package managers (with version) are defined in Dockerfile per ecosystem. For example: NPM and YARN ecosystem that uses NPM, defines the package manager as following in Dockerfile.

```
ARG NPM_VERSION=9.6.5
RUN corepack install npm@$NPM_VERSION --global
```

Once the docker container is up, the respective package managers (with mentioned version) are downloaded and installed within container. The per ecosystem file updater uses the package manager to run the updates. i.e.

```
def self.run_npm_command(command, fingerprint: command)
        SharedHelpers.run_shell_command("corepack npm #{command}", fingerprint: "corepack npm #{fingerprint}")
end
```
_**This static version enforcement poses some issues with updates.**_

a. As we are targeting version enforcement to maximize compatibility and thus dependabot is not always using latest version of package manager. A repo owner can specify a new version of package manager in his code manifest file (i.e. package.json). The compatibility mismatch can create issues while updating the dependencies.

b. A repo project may decide to lock the package manager version due to legacy nature of project. This means that package manager will refuse to update the dependences due to version mismatch.

**Returned error response example**

`Unsupported engine: wanted: {"node":">=16.16.0 <17.0.0 || ^18"} (current: {"node":"v20.16.0","pnpm":"9.4.0"})`

Here the node ver 20.16.0 is installed at docker container initial run (static version mentioned in _DockerFile_). As the manifest file has mentioned an alternate version range (16.16.0 <17.0.0 || ^18), native helper will refuse to update the dependencies due to version mismatch.

**Solution:** A proposed solution is to install a version of package manager at runtime that is compatible with one mentioned in manifest file. The manifest file can be parsed (i.e. "engines": {"pnpm": "9.1.2"}) and a compatible version matching the spec can be installed at runtime prior to running updates. runtime can be installed using corepack (`corepack install pnpm@9.1.2 --global`).

### The flow description is as following:

**File fetching:** During file fetching stage, the manifest files in repo are parsed for setting up any prerequisite for _file updater_ stage. Npm and Yarn manifest file (package.json) contains **engines** tag that specify the package manager (i.e. npm, yarn, pnpm) along with version number (i.e. "2.3.2"). _engines_ is parsed and sent as version for install using _corepack_ command.

```
SharedHelpers.run_shell_command(
          "corepack install #{name}@#{version} --global --cache-only",
          fingerprint: "corepack install <name>@<version> --global --cache-only"
        ) 
```

As for now for limited introduction of this feature. We are only allowing versions that match a linear strict form, such as v20.2.3 .  Other tag variations include `"yarn": ">=1.22",  "node": ">=18",  "node": "^18",  "node":">=16.16.0 <17.0.0 || ^18"` which will not be allowed at this time.

Manifest files may also include tag `"packageManager"` which specify which package manager (i.e. pnpm, yarn) is intended to be used for updates. A scenario might arise where `packageManager` has version specificity such as `"packageManager": "yarn@3.6.3"`, incase there is a conflict in version specs between `engines` and `packageManager` such as `"engines": {"yarn": ">=1.22"}` and `"packageManager": "yarn@3.6.3"` , "packageManager" setting will be prioritized.

**File updating:** Once the runtime engines package manager version is downloaded, the update is run as usual. A matching runtime environment will be available and updates can be run.

### Anything you want to highlight for special attention from reviewers?

Please provide your valuable feedback and suggestions.

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

**Comment by sachin-sandhu** [comment]
**Background:** Dependabot relies on ecosystem helpers (package managers) to perform dependencies updates. The package managers (with version) are defined in Dockerfile per ecosystem. For example: NPM and YARN ecosystem that uses NPM, defines the package manager as following in Dockerfile.

```
ARG NPM_VERSION=9.6.5
RUN corepack install npm@$NPM_VERSION --global
```

Once the docker container is up, the respective package managers (with mentioned version) are downloaded and installed within container. The per ecosystem file updater uses the package manager to run the updates. i.e.

```
def self.run_npm_command(command, fingerprint: command)
        SharedHelpers.run_shell_command("corepack npm #{command}", fingerprint: "corepack npm #{fingerprint}")
end
```
_**This static version enforcement poses some issues with updates.**_

a. As we are targeting version enforcement to maximize compatibility and thus dependabot is not always using latest version of package manager. A repo owner can specify a new version of package manager in his code manifest file (i.e. package.json). The compatibility mismatch can create issues while updating the dependencies.

b. A repo project may decide to lock the package manager version due to legacy nature of project. This means that package manager will refuse to update the dependences due to version mismatch.

**Returned error response example**

`Unsupported engine: wanted: {"node":">=16.16.0 <17.0.0 || ^18"} (current: {"node":"v20.16.0","pnpm":"9.4.0"})`

Here the node ver 20.16.0 is installed at docker container initial run (static version mentioned in _DockerFile_). As the manifest file has mentioned an alternate version range (16.16.0 <17.0.0 || ^18), native helper will refuse to update the dependencies due to version mismatch.

**Solution:** A proposed solution is to install a version of package manager at runtime that is compatible with one mentioned in manifest file. The manifest file can be parsed (i.e. "engines": {"pnpm": "9.1.2"}) and a compatible version matching the spec can be installed at runtime prior to running updates. runtime can be installed using corepack (`corepack install pnpm@9.1.2 --global`).

### The flow description is as following:

**File fetching:** During file fetching stage, the manifest files in repo are parsed for setting up any prerequisite for _file updater_ stage. Npm and Yarn manifest file (package.json) contains **engines** tag that specify the package manager (i.e. npm, yarn, pnpm) along with version number (i.e. "2.3.2"). _engines_ is parsed and sent as version for install using _corepack_ command.

```
SharedHelpers.run_shell_command(
          "corepack install #{name}@#{version} --global --cache-only",
          fingerprint: "corepack install <name>@<version> --global --cache-only"
        ) 
```

As for now for limited introduction of this feature. We are only allowing versions that match a linear strict form, such as v20.2.3 .  Other tag variations include `"yarn": ">=1.22",  "node": ">=18",  "node": "^18",  "node":">=16.16.0 <17.0.0 || ^18"` which will not be allowed at this time.

Manifest files may also include tag `"packageManager"` which specify which package manager (i.e. pnpm, yarn) is intended to be used for updates. A scenario might arise where `packageManager` has version specificity such as `"packageManager": "yarn@3.6.3"`, incase there is a conflict in version specs between `engines` and `packageManager` such as `"engines": {"yarn": ">=1.22"}` and `"packageManager": "yarn@3.6.3"` , "packageManager" setting will be prioritized.

**File updating:** Once the runtime engines package manager version is downloaded, the update is run as usual. A matching runtime environment will be available and updates can be run.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (kbukum1) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:94]: I think there is `linting` problem here. Reference: https://github.com/dependabot/dependabot-core/actions/runs/10627392075/job/29460440977?pr=10510

File: npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb
Comment (kbukum1) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb:29]: You can add `?` at the end of the function to show it is boolean check function. That's increase readability.

`valid_extracted_version?`

**Review by kbukum1 — COMMENTED** [review]
Is it possible to add Sorbet typings if it doesn't cause issues?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb]: fixed

File: npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb
Comment (abdulapopoola) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb:37]:
```suggestion
      def valid_extracted_version?(version)
         version.match?(NODE_ENGINE_SUPPORTED_REGEX)
```

Comment (abdulapopoola) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb:37]:
```
      def valid_extracted_version?(version)
        return true if version.match?(NODE_ENGINE_SUPPORTED_REGEX)

        false
      end
```
is same as 
```
      def valid_extracted_version?(version)
        version.match?(NODE_ENGINE_SUPPORTED_REGEX)
      end
```

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (abdulapopoola) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:95]: Any reason for this regex change?

**Review by abdulapopoola — APPROVED** [review]
Approving pending comment resolution and signoff by Jamie. Let's do a careful rollout and validation in production to be sure there are no regressions and this works as expected.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:95]: @abdulapopoola , there was a match issue between npm and pnpm. "p**npm**" has substring "npm" . So made the change to match exact engine names.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb:35]: @abdulapopoola , fixed it , thanks for suggestion.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (JamieMagee) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:27]: nit: this is hard to read, but I think it can be simplified:

```suggestion
          unless @package_manager&.start_with?("#{name}@") || @package_manager == name.to_s)
```

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (JamieMagee) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:48]: nit: same as above, use the safe navigation operator

```suggestion
          return unless @package_manager&.start_with?("#{name}@")
```

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (JamieMagee) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:51]:
```suggestion
        version ||= requested_version(name)
```

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (JamieMagee) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:112]:
```suggestion
        if engine_versions&.empty?
```

File: npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb
Comment (JamieMagee) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb:11]: This doesn't quite match your comment above. It will match `20` and `20.0.0`, but not `20.0`.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb
Comment (JamieMagee) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb:1]:
```suggestion
# typed: strict
```

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (yeikel) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:22]:
```suggestion
        # we go for the specificity mentioned in packageManager (6.0.2)
```

Comment (yeikel) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:22]: This wording is a bit strange(to me) but there is a typo

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (yeikel) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:23]: This comment does not add a lot a value. Can it just be removed?

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb]: @JamieMagee , fixed

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb]: @JamieMagee , fixed it

File: npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb]: @JamieMagee , thanks for noticing , fixed.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:50]: @JamieMagee , i think _nil?_ check needs exclusive condition check. as this condition did already exist (previous dev work), i have not tried to alter it (i tried and it was failing test cases) . I think we can leave it as it is for now.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb]: @yeikel fixed it.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb]: @JamieMagee , i have kept _nil?_ check for regression (makes easier to revert back in case). I have applied safe nav operators.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb
Comment (abdulapopoola) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:21]: can we please do a PR to add Sorbet typing to this file as we agreed some weeks ago (this can be a separate PR, cc: @kbukum1 )

Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb:21]: @abdulapopoola , ill create a separate PR for this. (i added sorbet for new functions already)

File: npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb
Comment (JamieMagee) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/version_selector.rb:16]: I think you can be more specific here

```suggestion
      sig { params(manifest_json: T::Hash[String, String], name: String).returns(T::Hash[Symbol, T.untyped]) }
```

**Review by JamieMagee — APPROVED** [review]
Approved pending changes

**Pull request merged**
